google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentRuntimeImpl.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (C) 2016 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.cloud;
+
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.gson.Gson;
+import com.google.gson.JsonParseException;
+import com.google.gson.reflect.TypeToken;
+
+import com.intellij.icons.AllIcons.General;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import com.intellij.remoteServer.runtime.deployment.DeploymentRuntime;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.Type;
+import java.util.List;
+
+import javax.swing.SwingUtilities;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+/**
+ * Takes care of undeploy and stop of App Engine applications.
+ */
+class AppEngineDeploymentRuntimeImpl extends DeploymentRuntime {
+
+  private static final String STOP_CONFIRMATION_URI_OPEN_TAG =
+      "<a href='https://cloud.google.com/appengine/docs/java/console/#versions'>";
+  private static final String STOP_CONFIRMATION_URI_CLOSE_TAG = "</a>";
+
+  private Project project;
+  private AppEngineHelper appEngineHelper;
+  private LoggingHandler loggingHandler;
+  private DeployOutput deployOutput;
+
+  public AppEngineDeploymentRuntimeImpl(
+      @NotNull Project project,
+      @NotNull AppEngineHelper appEngineHelper,
+      @NotNull LoggingHandler loggingHandler,
+      @NotNull String deploymentOutput) {
+    this.project = project;
+    this.appEngineHelper = appEngineHelper;
+    this.loggingHandler = loggingHandler;
+
+    try {
+      deployOutput = parseDeployOutputToService(deploymentOutput);
+    } catch (JsonParseException e) {
+      loggingHandler.print(
+          GctBundle.message("appengine.deployment.version.extraction.failure") + "\n");
+      Logger.getInstance(AppEngineDeploymentRuntimeImpl.class).warn(
+          "Could not retrieve service/version info of deployed application", e);
+      deployOutput = null;
```

Comment:
One improvement would be to move the parsing to the DeployAction. Then if its successful construct the Runtime with undeploy enabled (and disabled if not). 

We still have that useless Runtime but thats the pattern that IJ uses: even if undeploy was not supported at all we still have to construct it and override `isUndeploySupported` because of the callback api.